lib/shared/sdk/usbboot/index.js

```diff
@@ -0,0 +1,318 @@
+          debug(`Getting the size of ${fileMessage.fileName}`)
+          const fileSize = _.get(files, [ fileMessage.fileName, 'length' ])
+
+          if (_.isNil(fileSize)) {
+            debug(`Couldn't find ${fileMessage.fileName}`)
+            debug('Sending error signal')
+            return protocol.sendErrorSignal(device)
+          }
+
+          debug(`Sending size: ${fileSize}`)
+          return protocol.sendBufferSize(device, fileSize)
+        }
+
+        if (fileMessage.command === protocol.FILE_MESSAGE_COMMAND_READ_FILE) {
+          debug(`Reading ${fileMessage.fileName}`)
+          const fileBuffer = _.get(files, [ fileMessage.fileName ])
+
+          if (_.isNil(fileBuffer)) {
+            debug(`Couldn't find ${fileMessage.fileName}`)
+            debug('Sending error signal')
+            return protocol.sendErrorSignal(device)
+          }
+
+          return protocol.write(device, endpoint, fileBuffer)
+        }
+
+        return Bluebird.reject(new Error(`Unrecognized command: ${fileMessage.command}`))
+      }).then(() => {
+        debug('Starting again')
+        return startFileServer(device, endpoint, files)
+      })
+    })
+}
+
+exports.flash = (device, options = {}) => {
+  // We need to open the device in order to access _configDescriptor
+  debug(`Opening device: ${device.name}`)
+  device.open()
+
+  // Handle 2837 where it can start with two interfaces, the first
+  // is mass storage the second is the vendor interface for programming
+  const addresses = {}
+  /* eslint-disable no-underscore-dangle */
+  if (device._configDescriptor.bNumInterfaces === USB_ENDPOINT_INTERFACES_SOC_BCM2835) {
+  /* eslint-enable no-underscore-dangle */
+    addresses.interface = 0
+    addresses.endpoint = 1
+  } else {
+    addresses.interface = 1
+    addresses.endpoint = 3
+  }
+
+  const deviceInterface = device.interface(addresses.interface)
+  debug(`Claiming interface: ${addresses.interface}`)
+  deviceInterface.claim()
+
+  const endpoint = deviceInterface.endpoint(addresses.endpoint)
+
+  const serialNumberIndex = device.deviceDescriptor.iSerialNumber
+  debug(`Serial number index: ${serialNumberIndex}`)
+
+  if (serialNumberIndex === USB_DESCRIPTOR_NULL_INDEX) {
+    return writeBootCode(device, endpoint, _.get(options.files, [ 'bootcode.bin' ]))
+      .finally(_.partial(closeDevice, device))
+      .delay(USB_OPERATION_DELAY_MS).then(() => {
+        debug('Waiting for device to come back')
+        return waitForDevice({
+          deviceDescriptor: {
+            idVendor: USB_VENDOR_ID_BROADCOM_CORPORATION,
+            idProduct: 0x2764,
+            iSerialNumber: 1
+          }
+        }, {
+          retries: 5,
+          delay: 1000
+        })
+      }).then((newDevice) => {
+        if (_.isNil(newDevice)) {
+          throw new Error(`Device ${device.name} never came back`)
+        }
+
+        return exports.flash(newDevice, options)
```

Comment:
Hmmm, so (unless I'm misunderstanding) this is a recursive function? Is that right?